internal/cluster/cluster.go

```diff
@@ -235,6 +240,101 @@ func (m *Manager) RefreshStatus(ctx context.Context, c *common.Cluster, db *gorm
+	exists, err := m.objectHandler.DoesObjectExist(ctx, objectName)
+	if err != nil {
+		log.WithError(err).Warnf("Failed to get object status")
+		return false
+	}
+
+	if !exists {
+		return false
+	}
+
+	status, reason := m.getClusterMonitoringOperatorsStatus(cluster)
+	log.Infof("Monitoring status: %s with reason: %s", status, reason)
+
+	return funk.Contains([]models.OperatorStatus{models.OperatorStatusAvailable, models.OperatorStatusFailed}, status)
+}
+
+func (m *Manager) getClusterMonitoringOperatorsStatus(cluster *common.Cluster) (models.OperatorStatus, string) {
+	numberOfAvailableOperators := 0
+
+	for _, operator := range cluster.MonitoredOperators {
+		switch operator.Status {
+		case models.OperatorStatusFailed:
+			return models.OperatorStatusFailed, fmt.Sprintf("Operators %s has failed with info %s", operator.Name, operator.StatusInfo)
+		case models.OperatorStatusAvailable:
+			numberOfAvailableOperators++
+		case models.OperatorStatusProgressing:
+			continue
+		}
+	}
+
+	if numberOfAvailableOperators == len(cluster.MonitoredOperators) {
+		return models.OperatorStatusAvailable, fmt.Sprintf("All %d monitored operators are available", len(cluster.MonitoredOperators))
+	}
+
+	return models.OperatorStatusProgressing, fmt.Sprintf("Not all monitored operators have finished %d/%d", numberOfAvailableOperators, len(cluster.MonitoredOperators))
+}
+
+func (m *Manager) CompleteInstallation(ctx context.Context, c *common.Cluster, successfullyFinished bool, reason string) error {
+	log := logutil.FromContext(ctx, m.log)
+
+	if swag.StringValue(c.Status) == models.ClusterStatusInstalled {
+		log.Infof("Cluster %s installation was already completed", *c.ID)
+		return nil
+	}
+
+	err := m.sm.Run(TransitionTypeCompleteInstallation, newStateCluster(c), &TransitionArgsCompleteInstallation{
+		ctx:       ctx,
+		isSuccess: successfullyFinished,
+		reason:    reason,
+	})
+	if err != nil {
+		return common.NewApiError(http.StatusConflict, err)
+	}
+
+	return m.updateCompletion(ctx, log, c, successfullyFinished, reason)
+}
+
+func (m *Manager) updateCompletion(ctx context.Context, log logrus.FieldLogger, cluster *common.Cluster, successfullyFinished bool, reason string) error {
+	result := models.ClusterStatusInstalled
+
+	severity := models.EventSeverityInfo
+	eventMsg := fmt.Sprintf("Successfully finished installing cluster %s", cluster.Name)
+	if !successfullyFinished {
+		result = models.ClusterStatusError
+		severity = models.EventSeverityCritical
+		eventMsg = fmt.Sprintf("Failed installing cluster %s. Reason: %s", cluster.Name, reason)
+	}
+
+	m.metricAPI.ClusterInstallationFinished(log, result, cluster.OpenshiftVersion,
+		*cluster.ID, cluster.EmailDomain, cluster.InstallStartedAt)
+	m.eventsHandler.AddEvent(ctx, *cluster.ID, nil, severity, eventMsg, time.Now())
+
+	if m.ocmClient != nil {
+		_, err := m.ocmClient.AccountsMgmt.UpdateSubscription(ctx, cluster.AmsSubscriptionID, cluster.OpenshiftClusterID)
```

Comment:
if UpdateSubscription nobody will retry to update it one more time.
Right now controller will retry it for 2 hours cause it will get failure from complete installation. Is it ok not to UpdateSubscription ?
@ybettan

Reply:
Umm technically no, we must update it.
Not updating it means that:
1. the user don't know his cluster is ready
2. No one will "own" the subscription (reserved subscription are owned by the service and active one by AMS

The way it work today, we failed the installation if we failed to update the subscription.